### PR TITLE
Add new DSL statement: def_type & ref_type

### PR DESCRIPTION
**Pull request by mohanson**
Add two rules for preprocessing of tx_file:

- replace `{{ def_type ANY_NAME }}` to `{
            "code_hash": "0x00000000000000000000000000000000000000000000000000545950455f4944",
            "hash_type": "type",
            "args": "hex(ANY_NAME)"
          }`
- replace `{{ ref_type ANY_NAME }}` to hex data calculated by `script_hash(type_id_ANY_NAME)`

### CODE REVIEW

File: ckb-debugger/src/main.rs [hunk 1]
```diff
@@ -173,24 +177,50 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                std::fs::read(path).unwrap()
             };
-            let data = std::fs::read(path).unwrap();
             if cap1 == "data" {
-                return format!("0x{}", hex::encode(data));
+                return format!("0x{}", hex::encode(read_file(cap2)));
             }
             if cap1 == "hash" {
-                return format!("0x{}", hex::encode(&blake2b_256(data)));
+                return format!("0x{}", hex::encode(&blake2b_256(read_file(cap2))));
```
Comment (XuJiandong) [inline ckb-debugger/src/main.rs]: Remove "0x". We need following combination: "0xFFFF{{ hash xxx}}"

File: ckb-debugger/src/main.rs [hunk 2]
```diff
@@ -173,24 +177,50 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+                    root.join(cap2)
+                } else {
+                    Path::new(cap2).to_path_buf()
+                };
+                std::fs::read(path).unwrap()
             };
-            let data = std::fs::read(path).unwrap();
             if cap1 == "data" {
-                return format!("0x{}", hex::encode(data));
+                return format!("0x{}", hex::encode(read_file(cap2)));
```
Comment (XuJiandong) [inline ckb-debugger/src/main.rs]: It's not necessary to remove this "0x" here. For the sake of consistency, it is better to remove "0x".

File: ckb-debugger/src/main.rs [hunk 3]
```diff
@@ -173,24 +177,50 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+        for caps in mock_re_def_type.captures_iter(&mock_tx) {
+            let type_id_name = &caps[1];
+            let type_id_script = Script::new_builder()
+                .args(Bytes::from(type_id_name.to_string()).pack())
+                .code_hash(TYPE_ID_CODE_HASH.pack())
+                .hash_type(ScriptHashType::Type.into())
+                .build();
+            let type_id_script_hash = type_id_script.calc_script_hash();
+            let type_id_script_hash = format!("{:x}", type_id_script_hash);
+            mock_type_id_dict.insert(type_id_name.to_string(), type_id_script_hash);
```
Comment (XuJiandong) [inline ckb-debugger/src/main.rs:195]: Report error if type_id_name is already existing.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: ckb-debugger/src/main.rs (hunk 3 above)
Comment (mohanson) [reply ckb-debugger/src/main.rs:195]: Fixed

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**